### DISCUSSION TITLE
About the mpirun

### DISCUSSION THREAD

**Discussion by Sduwzy** [chombo-discharge/chombo-discharge#465] [Q&A] [answered]
Hi,

I encountered an issue when running the compiled program using mpirun (running the DeterministicAir example).

In the 2D case, if I set the number of threads np>2, the program crashes.
In the 3D case, if I set the number of threads np>1, the program crashes.
The error message is as follows:

```
[yousatech-MZ73-LM0-000:433373] *** Process received signal ***
[yousatech-MZ73-LM0-000:433373] Signal: Segmentation fault (11)
[yousatech-MZ73-LM0-000:433373] Signal code: (128)
[yousatech-MZ73-LM0-000:433373] Failing at address: (nil)
......
mpirun noticed that process rank 1 with PID 0 on node yousatech-MZ73-LM0-000 exited on signal 11 (Segmentation fault).
```

My system configuration is:

CPU: AMD EPYC9754
Memory: 400GB
Could you please help me understand the cause of this issue?

Thank you!
![image](https://github.com/chombo-discharge/chombo-discharge/assets/99463183/fcccb137-3412-42b9-bccd-84bd27ac6ddf)

**Comment by rmrsk** [comment]
Before I try to reproduce; can you paste your Make.defs.local and contents of pout.0? That might tell us something about why this happened.

**Comment by Sduwzy** [comment]
Thank you for immediate response, the attachment is my Make.defs.local and contents of pout.0., in order to update in Github, I add [.txt] at the end of the file name.

[Make.defs.local.txt](https://github.com/user-attachments/files/15781400/Make.defs.local.txt)
[pout.0.txt](https://github.com/user-attachments/files/15781402/pout.0.txt)

**Comment by rmrsk** [comment]
Thanks for the files. First; in your screenshot you have both debug and optimization mode enabled. That's not wrong, but DEBUG mode enables a bunch of deep-level checks to assert code correctness in various places, and thus it tends to slow some operations down to a crawl. I would recommend that you set DEBUG to FALSE and OPT to HIGH.

One other thing is that the configuration file you attached can't be the same as the one you used to compile chombo-discharge, because it has DEBUG=FALSE but your executable has DEBUG=TRUE. 

All this being said, I could not reproduce the error that you saw. I did find another error in the input file that you used for the program, which used a very coarse mesh and a huge timestep so that geometric multigrid failed to converge in the semi-implicit solve. I'm issuing a  fix for that.

My guess is that your problem is due to a bad installation, e.g. from a few files not being recompiled when you change compilation settings. This can happen e.g. if memory tracking is disabled when you compile the source code, but enabled when you compile your program. There are other places where similar things can occur, and they are best avoided by recompiling the whole thing. 

Can you try to recompile? First, clean up everything by going to $DISCHARGE_HOME and type 'make pristine'. Then go to Exec/Examples/CdrPlasma/DeterministicAir and type 'make -s program'. Try to run the program from there.

**Comment by Sduwzy** [comment]
Thank you for your kind response.

I have done as you said, clean the program and recompile it in DeterministicAir. The problem still occured, when I tried to run it with
`mpirun -np 2 program2d.Linux.64.mpic++.gfortran.OPTHIGH.MPI.ex positive2d.inputs 
Everything goes well.
However, when I tried to increase [np], and run it as: 
`mpirun -np 3 program2d.Linux.64.mpic++.gfortran.OPTHIGH.MPI.ex positive2d.inputs `
The same error occurred, the message is shown as: 
![image](https://github.com/chombo-discharge/chombo-discharge/assets/99463183/5fb887c9-3821-4110-ae21-ddd59d864b49)


I have attached the pout0.txt here, could you please check it?

Thank you again for your help.


[pout0.txt](https://github.com/user-attachments/files/15785205/pout0.txt)

**Comment by rmrsk** [comment]
The output-file looks fine. As I can't reproduce, I don't know what is going on here. One way of moving forward is to set all the verbosity-related flags in the input file to a large number (e.g., 99). That will enable more verbose output into the progress-report files (pout.0, pout.1,...) so we can see _where_ things go bad. Are you also using HDF5 for output? And have you made sure that HDF5 is compiled with MPI and pointed the configuration file to the correct place?

**Comment by Sduwzy** [comment, marked as answer]
After some exploration, I found the cause of the issue: my Ubuntu version was 18, which only supports MPI version 2.0. This led to the problems I was experiencing. Once I upgraded my Ubuntu to version 20 and MPI to version 4.0 or higher, the issues were resolved. 

Thank you very much for your patient assistance.
